python/events/evaluation_tools/events/event_detection/decomposition.py

```diff
@@ -0,0 +1,168 @@
+        """
+    # Flip series values to run filter in reverse
+    if reverse:
+        series = pd.Series(series.values[::-1], 
+            index=series.index)
+
+    # Smooth series
+    smooth = series.ewm(halflife=halflife, times=series.index, 
+        adjust=False).mean()
+    
+    # Estimate a seasonal trend using a rolling minimum
+    trend = smooth.rolling(window=window).min()
+
+    # Remove the seasonal trend
+    detrended = smooth - trend
+
+    # Assume a residual equal to twice the detrended median
+    residual = detrended.median() * 2.0
+
+    # Remove the residual
+    detrended = detrended - residual
+
+    # Eliminate negative values
+    detrended[detrended < 0.0] = 0.0
+
+    # Restore reversed series
+    if reverse:
+        # Return detrended series
+        return pd.Series(detrended.values[::-1], 
+            index=detrended.index)
+
+    # Return detrended series
+    return detrended
+
+def mark_event_flows(series, halflife, window) -> pd.Series:
+    """Model the trend in a streamflow time series by taking the mean
+        of two rolling minimum filters applied in a forward and 
+        backward fashion. Remove the trend and residual components.
+        Return the boolean time series that indicates whether an 
+        individual value in the original streamflow time series
+        is part of an event (True) or not part of an event (False).
+        
+        Parameters
+        ----------
+        series: pandas.Series with a DateTimeIndex, required
+            The original streamflow time series.
+        halflife: float, str, timedelta, required
+            Specifies the decay term for `pandas.Series.ewm.mean`. This filter 
+            is used to smooth over interevent noise and reduce detection 
+            of normal instrument fluctuations as events.
+        window: int, offset, or BaseIndexer subclass, required
+            Size of the moving window for `pandas.Series.rolling.min`.
+            This filter is used to model the trend in `series`.
+            
+        Returns
+        -------
+        event_points: pandas.Series
+            Boolean time series where True indicates the associated value
+            in the `series` is part of an event.
+        
+        """
+    # Detrend with a forward filter
+    forward = detrend_streamflow(series, halflife, window)
+
+    # Detrend with a backward filter
+    backward = detrend_streamflow(series, halflife, window, True)
+
+    # Take the mean of the forward and backward trends
+    detrended = np.mean([forward, backward], axis=0)
+
+    # Assume a residual equal to twice the detrended median
+    residual = np.median(detrended) * 2.0
+
+    # Remove the residual
+    detrended = detrended - residual
+
+    # Eliminate negative values
+    detrended[detrended < 0.0] = 0.0
+
+    # Return mask of non-zero detrended flows
+    return pd.Series((detrended > 0.0), index=series.index)
+
+def list_events(series, halflife, window) -> pd.DataFrame:
```

Comment:
I don't think it would hurt to either link to the manuscript, so leave a TODO, and/or briefly mention the assumptions made by the method.

Reply:
Other than that, this module looks top class!

Reply:
Added explicit citation